### PR TITLE
Handle bad characters in field names

### PR DESCRIPTION
**Pull request by JasonStoltz**
closes #https://github.com/elastic/app-search-firestore-extension/issues/16

The linked issue and the tests below should be descriptive of the changes.

### CODE REVIEW

File: POSTINSTALL.md [hunk 1]
```diff
@@ -136,12 +136,12 @@ To reindex data, use the steps listed above for "Backfill or import existing doc
 
 The TLDR for this section is:
 
-- Field names will be lowercased before being indexed, so the field name in app search may differ from the field name in Firestore.
 - `text` and `number` type fields are indexed as-is to App Search.
 - `geo` and `timestamp` fields are formatted slightly differently when indexed.
 - `map`, `boolean`, and `reference` are not supported by App Search, they will be indexed as text.
 - nested arrays are not supported at all and will be dropped before indexing in App Search.
-- While `map`s are not supported, you _can_ specify that fields within a map get indexed as top level fields in App Search.
+- While `map`s are not supported, you _can_ specify that fields within a map get indexed as top level fields in App Search, using the `__` syntax when configuring indexed fields
```
Comment (byronhulcher) [inline POSTINSTALL.md:143]: Is using `__` over `.` to lower conflict risk with an actual field name that uses `.`?  Admittedly this syntax does feel a little unnatural.

File: POSTINSTALL.md [hunk 2]
```diff
@@ -136,12 +136,12 @@ To reindex data, use the steps listed above for "Backfill or import existing doc
...
+- App Search only supports lower-cased alphanumeric characters and underscores ("\_") in field names. This extension will rename fields that don't match, or you can use the `::` syntax to specify what it is renamed to when configuring indexed fields.
```
Comment (byronhulcher) [inline POSTINSTALL.md:144]: `::` is good 👍

File: functions/src/toAppSearch.ts [hunk 3]
```diff
@@ -19,11 +19,29 @@ export const toAppSearch = (
...
     if (fieldValue === undefined) return acc;
 
-    // App search doesn't support dot notation so we need to join them with "__"
-    const processedFieldName = fieldName.split(".").join("__").toLowerCase();
+    // App Search only supports lowercased alpha numeric names or underscores
+    const processedFieldName = (renameTo || parsedFieldName)
+      .replace(/[^A-Za-z0-9_]/g, "")
+      .toLowerCase();
+
+    if (processedFieldName === "") return acc;
```
Comment (byronhulcher) [inline functions/src/toAppSearch.ts:44]: I wonder if we should log something to say "hey we dropped this field" or "hey we dropped these fields:".

**Review by byronhulcher — APPROVED** [review]
This is good.  I had some questions but none of them represent blockers, this could be merged as is.

File: POSTINSTALL.md (hunk 1 above)
Comment (JasonStoltz) [reply POSTINSTALL.md:143]: I agree it feels a bit unnatural. It's to reduce the chance of conflicts *and* the name of the indexed field ends up matching the name of the field that actually gets indexed... one less point of confusion. Meaning, when you specify `field__subfield` it will actually be indexed as `field__subfield`

File: functions/src/toAppSearch.ts (hunk 3 above)
Comment (JasonStoltz) [reply functions/src/toAppSearch.ts:44]: That's a good idea.
